### PR TITLE
add base for config

### PR DESCRIPTION
**Pull request by xvjiarui**
This PR added `_base_` in Config for config file inheritance.

### CODE REVIEW

**Comment by hellock** [comment]
The desired configuration system should allow multiple bases (components).
 
## faster_rcnn_r101_fpn_1x_coco.py
```python
_base_: ['faster_rcnn_r50_fpn.py', 'coco.py', 'schedule_1x.py', 'default_runtime.py']
model = dict(
    pretrained='torchvision://resnet101',
    backbone=dict(depth=101)
)
work_dir = './work_dirs/faster_rcnn_r101_fpn_1x_coco'
```

### faster_rcnn_r50_fpn.py
```python
model = dict(
    type='FasterRCNN',
    pretrained='torchvision://resnet50',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        num_outs=5),
    rpn_head=dict(
        type='RPNHead',
        in_channels=256,
        feat_channels=256,
        anchor_scales=[8],
        anchor_ratios=[0.5, 1.0, 2.0],
        anchor_strides=[4, 8, 16, 32, 64],
        target_means=[.0, .0, .0, .0],
        target_stds=[1.0, 1.0, 1.0, 1.0],
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
        loss_bbox=dict(type='SmoothL1Loss', beta=1.0 / 9.0, loss_weight=1.0)),
    bbox_roi_extractor=dict(
        type='SingleRoIExtractor',
        roi_layer=dict(type='RoIAlign', out_size=7, sample_num=2),
        out_channels=256,
        featmap_strides=[4, 8, 16, 32]),
    bbox_head=dict(
        type='SharedFCBBoxHead',
        num_fcs=2,
        in_channels=256,
        fc_out_channels=1024,
        roi_feat_size=7,
        num_classes=81,
        target_means=[0., 0., 0., 0.],
        target_stds=[0.1, 0.1, 0.2, 0.2],
        reg_class_agnostic=False,
        loss_cls=dict(
            type='CrossEntropyLoss', use_sigmoid=False, loss_weight=1.0),
        loss_bbox=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0)))
# model training and testing settings
train_cfg = dict(
    rpn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.7,
            neg_iou_thr=0.3,
            min_pos_iou=0.3,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=256,
            pos_fraction=0.5,
            neg_pos_ub=-1,
            add_gt_as_proposals=False),
        allowed_border=0,
        pos_weight=-1,
        debug=False),
    rpn_proposal=dict(
        nms_across_levels=False,
        nms_pre=2000,
        nms_post=2000,
        max_num=2000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        assigner=dict(
            type='MaxIoUAssigner',
            pos_iou_thr=0.5,
            neg_iou_thr=0.5,
            min_pos_iou=0.5,
            ignore_iof_thr=-1),
        sampler=dict(
            type='RandomSampler',
            num=512,
            pos_fraction=0.25,
            neg_pos_ub=-1,
            add_gt_as_proposals=True),
        pos_weight=-1,
        debug=False))
test_cfg = dict(
    rpn=dict(
        nms_across_levels=False,
        nms_pre=1000,
        nms_post=1000,
        max_num=1000,
        nms_thr=0.7,
        min_bbox_size=0),
    rcnn=dict(
        score_thr=0.05, nms=dict(type='nms', iou_thr=0.5), max_per_img=100)
    # soft-nms is also supported for rcnn testing
    # e.g., nms=dict(type='soft_nms', iou_thr=0.5, min_score=0.05)
)
```
### coco.py
```python
dataset_type = 'CocoDataset'
data_root = 'data/coco/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(1333, 800), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
]
test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1333, 800),
        flip=False,
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='RandomFlip'),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),
        ])
]
data = dict(
    imgs_per_gpu=2,
    workers_per_gpu=2,
    train=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/instances_train2017.json',
        img_prefix=data_root + 'train2017/',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/instances_val2017.json',
        img_prefix=data_root + 'val2017/',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=data_root + 'annotations/instances_val2017.json',
        img_prefix=data_root + 'val2017/',
        pipeline=test_pipeline))
evaluation = dict(interval=1, metric='bbox')
```

### schedule_1x.py
```python
# optimizer
optimizer = dict(type='SGD', lr=0.02, momentum=0.9, weight_decay=0.0001)
optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=1.0 / 3,
    step=[8, 11])
total_epochs = 12
```

### default_runtime.py
```python
# yapf:disable
log_config = dict(
    interval=50,
    hooks=[
        dict(type='TextLoggerHook'),
        # dict(type='TensorboardLoggerHook')
    ])
# yapf:enable
dist_params = dict(backend='nccl')
log_level = 'INFO'
load_from = None
resume_from = None
workflow = [('train', 1)]
```

**Comment by xvjiarui** [comment]
Good idea!

File: mmcv/utils/config.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ class Config(object):
     """
 
     @staticmethod
-    def fromfile(filename):
+    def file2dict(filename):
```
Comment (hellock) [inline mmcv/utils/config.py]: `_file2dict` for inner usage.

File: mmcv/utils/config.py [hunk 2]
```diff
@@ -117,13 +122,40 @@ def __init__(self, cfg_dict=None, filename=None):
...
+                base_filename)
+
+            def merge_a_into_b(a, b):
+                # merge dict a into dict b. values in a will overwrite b.
+                for k, v in a.items():
+                    if isinstance(v, dict) and k in b:
+                        assert isinstance(
+                            b[k],
+                            dict), "Cannot inherit key '{}' from base!".format(
+                                k)
```
Comment (hellock) [inline mmcv/utils/config.py]: The format is a little ugly. We can use an if-statement to raise exceptions.

File: mmcv/utils/config.py [hunk 3]
```diff
@@ -158,3 +214,29 @@ def __setitem__(self, name, value):
...
     def __iter__(self):
         return iter(self._cfg_dict)
+
+    def dump(self):
+        cfg_dict = super(Config, self).__getattribute__('_cfg_dict')
+        format_text = json.dumps(cfg_dict, indent=2)
+        return format_text
+
+    def merge_from_options(self, options):
+        """ Merge list into cfg_dict
```
Comment (hellock) [inline mmcv/utils/config.py]: Add a blank line here.

File: mmcv/utils/config.py [hunk 4]
```diff
@@ -158,3 +214,29 @@ def __setitem__(self, name, value):
...
+        Merge the dict parsed by MultipleKVAction into this cfg.
+        E.g., `options = ['model.backbone.depth', 50]`.
```
Comment (hellock) [inline mmcv/utils/config.py]: This example does not match with the api.

File: tests/data/config/l.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+_base_ = ['./l1.py', './l2.py', './l3.py']
```
Comment (hellock) [inline tests/data/config/l.py]: May be harder: `[./l1.py, ./l2.yaml, ./l3.json]`.

Comment (hellock) [inline tests/data/config/l.py]: May be harder: `[./l1.py, ./l2.yaml, ./l3.json]`.

**Pull request merged**